### DISCUSSION TITLE
DSL to JSON Conversion SDK

### DISCUSSION THREAD

**Discussion by dblclik** [openfga/community#13] [Ideas]
Currently, the only method for converting DSL to JSON is via the [Syntax Transformer](https://github.com/openfga/syntax-transformer/blob/main/src/friendly-to-api.ts).  While this works nicely for front end environments like the Playground, it requires teams either package the ts methods as CLI or host the transformers as APIs and write glue code.

A colleague and I are considering writing a library to make this simple to do directly in Rust, Go, Python, and JS (with other languages easy to add and would welcome collaboration).  Does anyone else feel like this is a tool they'd benefit from?  Having a custom low-level parser, AST, and converter could also make it easier to support DSL in the engine if we choose, and enable more tools to be built around the DSL w/o requiring integrating the typescript transformer library.

**Comment by maxmindlin** [comment]
We took a first pass at this and have something usable to show. [This repo](https://github.com/maxmindlin/openfga-dsl-parser) currently parses DSL document strings into an AST and then provides a transformer for serializing to JSON. It seems to work for many cases, as can be seen in unit and integration tests. It currently does limited syntax checking - it really only throws parser errors related to an inability to build the AST.

I also threw together a [Python binding library](https://github.com/maxmindlin/openfga-dsl-parser-python) that simply wraps the former library. It is installed via Pypi with no install requirements other than Python + pip on Linux, Windows, and MacOS. The neat part of this approach is that binding libraries like this (any language that can communicate via C FFI) can share the same core library while still having a native API.

If anybody has thoughts or suggestions please let me know!

**Comment by jon-whit** [comment]
@dblclik @maxmindlin this is awesome! Thanks for the support on this. We've been wanting to bring more native DSL to JSON transformation to clients across various languages. In the past we had discussed making a DSL to JSON transformer as a [WASM](https://webassembly.org/) module so that it's usable across a variety of languages (any of which that support the WASM runtime). This would allow us to use the WASM module in the browser for the Playground but also use it in OpenFGA (Go) directly.

Have you guys considered this approach at all? I think this would make it more cross-platform consumable. WDYT?

**Reply by maxmindlin** [reply]
@jon-whit Great to hear, thanks! Yes, I've considered wasm. The downside of using wasm across everything (IMO) is that a.) you have to have support for a wasm runtime (every modern language has support for a C FFI whereas wasm is still relatively young) and b.) it can be less performant than other options. The good thing about using Rust as a base for the core library is that it has fantastic support for the C FFI, as well as fantastic support for WASM via [WASM-bindgen](https://github.com/rustwasm/wasm-bindgen). In my other projects I have used wrapped the same core in C FFIs for server-side langs (python, nodejs, go, java) and then generated a wasm library for frontend. You could also use the wasm library in your nodejs, but I find it to be a little more cumbersome than just doing a lib using node addon files. The upside to wasm (like you said) is that wasm can be pretty platform agnostic, but I find github's OS matrices make generating the platform support pretty easy.

**Reply by dblclik** [reply]
@jon-whit - echoing everything Max said, a lot of it depends on what that target is. Being able to subsequently embed the output of what we are doing into a WASM offering is a great benefit, without requiring that every other integration support and be able to interface with WASM.  Once we've got more of the client libraries done, we can absolutely show how easy it is to integrate into WASM if that's something that there's an interest in 🙌

**Reply by jon-whit** [reply]
I'm definitely not opposed, but I think we should experiment with both options and see how we like them both. My fear with the C FFI approach is the ugly dependency chain that it brings with it with the C libraries (`libc` or `glibc`)  and the complexity these can bring to certain environments. libc and glibc are notorious for not being easy to integrate in certain container runtimes (alpine and similar). So if we can avoid linking in libc or glibc and invoking remote functions across those I think it would be worth it. I'm just starting to cry even thinking about my past experience with remote execution like this (Cython was especially painful and this feels similar to that) 😨

I think the closer we can get to a native binary format supported across multiple runtimes, the better.

**Reply by maxmindlin** [reply]
@jon-whit fair points! It would be pretty straight forward for me to throw together a library that compiles this to wasm. I have not, however, called into a wasm runtime from anything but JS.

**Reply by jon-whit** [reply]
@maxmindlin Let's do this... If you can get me a WASM module that integrates this library you've put together and provide an example of the WASM integration in JS, then I'll build around it to demonstrate an integration in Go as well. That's our primary use case right now. We want to be able to use the WASM module in JS for our frontend and use the WASM module in Go in our server to (potentially) allow clients to provide us DSL natively in the OpenFGA API going forward (instead of having to convert it to JSON). If our API could accept the DSL then that would eliminate the need to have cross language support (other than JS and Go) for the DSL transformer altogether.

That'll be a good line of work that will help us move things forward in a positive direction. How does that sound?

**Reply by maxmindlin** [reply]
@jon-whit sounds good to me! Thanks!

**Reply by jon-whit** [reply]
Thank you! It's awesome you're working on this, because this is something that's been on our radar for a hot minute but just haven't had time to get around to it quite yet. Thanks for the support 🚀

**Reply by dblclik** [reply]
@jon-whit - finally, someone who can relate with me to how awful Cython was lol

To be clear, I'm actually proposing that we offer both. Creating a WASM target and/or integrating the libraries that we're building into a WASM binary should be fairly trivial, but providing C FFI integration with all of the language targets we can would be less trivial (err, not trivial lol) if we started from the WASM side.

**Reply by dblclik** [reply]
Hey @jon-whit - We've gotten a WASM binary that builds from our Rust library [HERE](https://github.com/dblclik/openfga-dsl-parser-wasm), and I've created a pre-production release with the WASM binary included if you don't want to set up the rust toolchain to build your own ([LINK](https://github.com/dblclik/openfga-dsl-parser-wasm/releases/tag/v.0.1.0)).  We've tested this with JS and Go and it seems to work well 🤞 🎉 

If you check out the [Go example](https://github.com/dblclik/openfga-dsl-parser-wasm/blob/main/examples/go/main.go) I've created, you can see how I'm currently using the WASM binary w/ wasmer.  Let us know what you think!

**Reply by jon-whit** [reply]
@dblclik awesome! I'll take a look at your work this week and run it by the team and get back with you. Thanks 👍 !

**Reply by aaguiarz** [reply]
Hi all! 

I was thinking about the use case for this. 

Let's assume we have a CLI that can accept the model, validate it, transform it to JSON, and potentially post it to OpenFGA. 

Why would you need to have the DSL->JSON transformation on each SDK?

**Comment by aaguiarz** [comment]
The [CLI](https://github.com/openfga/cli?tab=readme-ov-file#transform-an-authorization-model) has now full support for transforming models.

There are also modules for Go and Javascript in the [Language](https://github.com/openfga/language/blob/main/README.md#transformer) repository to implement transformations in code. Java support is coming.
